### DISCUSSION TITLE
serialize char array in struct

### DISCUSSION THREAD

**Discussion by ahao1995** [nlohmann/json#3807] [Q&A]
```
struct person
{
    std::string name;
    std::string address;
    int         age;
    char        buf[1024];
};
NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(person, name, address, age, buf)

int main()
{
    person p;
    p.name    = "ahao";
    p.address = "shanghai";
    p.age     = 18;
    strcpy(p.buf, "test_buf");
    nlohmann::json j = p;
    person p2 = j.get<person>(); //error [json.exception.type_error.304] cannot use at() with string
}
```
when I use  ` j.get<person>();` I got a error, what should I do?
thank you!

**Comment by nlohmann** [comment]
This is an asymmetry between to_json and from_json for raw C-style char arrays. `char buf[1024]` matches the to_json overload for anything constructible into std::string, so `j["buf"]` ends up as a JSON *string*. But the from_json overload for fixed-size C arrays expects the source to be a JSON *array* and iterates it element-by-element with at(), which fails on a string value — hence 'cannot use at() with string'. There's no supported round-trip for raw char buffers via NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE. The fix is to avoid `char buf[1024]` and use `std::string` (or `std::array<char, 1024>`/a binary type) instead, or write a manual to_json/from_json pair for `person` that explicitly handles the buffer.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
